# <center>**Podstawy Sztucznej Inteligencji**</center>
## <center>Projekt 2</center>
### <center> Drzewa decyzyjne oraz sieci neuronowe przy predykcji spalonych obszarów lasu</center>



<center>Adrian Nadratowski</center>
<center>Marcin Skrzypkowski</center>

Celem projektu jest stworzenie modeli przewidujących spalony obszar lasu przy wykorzystaniu udostępnionych nam zestawów danych, pobranych z __Machine Learning Repository__ (https://archive.ics.uci.edu/ml/datasets/Forest+Fires). Naszym zadaniem było przeanalizować powyższe dane, a następnie zaprojektować, nauczyć i zwalidować dwa regresory: model sieci neuronowej oraz drzewa decyzyjnego. Sprawdzenie miało odbyć się z pomocą _k_-krotnej walidacji krzyżowej.

Biblioteki wykorzystane do projektu:
  * pandas
  * tensorflow
  * scikit-learn
  * matplotlib

# %TODO:
### do uzupełnienia:
  * kluczowe decyzje projektowe
  * opis struktury programu
  * instrukcja użytkownika (+opis interfejsu)
  * wkład poszczególnych autorów
  * analiza zbioru danych (również odnośnie brakujących wartości)
  

### Opis struktury programu

Projekt podzielony został na __X__ części. W pierwszej następuje analiza danych oraz sprawdzenie czy nie brajkuje jakichś wartości, które mogłyby wpłynąć na jakość uczenia. Następnie przedstawiona została implementacja drzew decyzyjnych oraz walidacja kilku modeli o różnych parametrach. Kolejnym punktem jest sieć neuronowa oraz jej walidacja. Na sam koniec obydwa sposoby modelowania zostały ze sobą porównane. 

### Instrukcja obsługi użytkownika

  * po uruchomieniu niniejszego projektu należy upewnić się, czy zainstalowane są wszystkie wymagane biblioteki wymienione we wprowadzeniu
  * w folderze _../data_ powinien znajdować się zbiór danych, na którym operować będzie poniższy skrypt
  * projekt składa się z __4__ głównych części: 
    1. import i walidacja danych
    2. model drzewa decyzyjnego
    3. model sieci neuronowej
    4. podsumowanie
    i w tej kolejności należy uruchamiać odpowiednie skrypty
  * każda część projektu uruchamiana jest kolejno od najwyższych do najniższych komórek
  * każda komórka jest szczegółowo opisana z wyjaśnieniem procesu w niej zaimplementowanego

### Analiza danych
Poniżej następuje import bibliotek niezbędnych do reprezentacji danych, ich import i reprezentacja

In [7]:
import pandas as pd
import numpy as np
import matplotlib as mpt
import math
%matplotlib inline

In [8]:
data = pd.read_csv('../data/forestfires.csv')
data

X  Y month  day  ...  RH  wind  rain   area
0    7  5   mar  fri  ...  51   6.7   0.0   0.00
1    7  4   oct  tue  ...  33   0.9   0.0   0.00
2    7  4   oct  sat  ...  33   1.3   0.0   0.00
3    8  6   mar  fri  ...  97   4.0   0.2   0.00
4    8  6   mar  sun  ...  99   1.8   0.0   0.00
5    8  6   aug  sun  ...  29   5.4   0.0   0.00
6    8  6   aug  mon  ...  27   3.1   0.0   0.00
7    8  6   aug  mon  ...  86   2.2   0.0   0.00
8    8  6   sep  tue  ...  63   5.4   0.0   0.00
9    7  5   sep  sat  ...  40   4.0   0.0   0.00
10   7  5   sep  sat  ...  51   7.2   0.0   0.00
11   7  5   sep  sat  ...  38   4.0   0.0   0.00
12   6  5   aug  fri  ...  72   6.7   0.0   0.00
13   6  5   sep  mon  ...  42   2.2   0.0   0.00
14   6  5   sep  wed  ...  21   4.5   0.0   0.00
15   6  5   sep  fri  ...  44   5.4   0.0   0.00
16   5  5   mar  sat  ...  27   5.4   0.0   0.00
17   8  5   oct  mon  ...  47   4.9   0.0   0.00
18   6  4   mar  wed  ...  35   4.0   0.0   0.00
19   6  4   apr  sat  ...  44   4.5   0.0   0.00
20   6  4   sep  tue  ...  40   2.7   0.0   0.00
21   5  4   sep  mon  ...  38   2.7   0.0   0.00
22   7  4   jun  sun  ...  44   4.5   0.0   0.00
23   7  4   aug  sat  ...  43   5.8   0.0   0.00
24   7  4   aug  sat  ...  32   5.8   0.0   0.00
25   7  4   aug  sun  ...  60   5.4   0.0   0.00
26   7  4   sep  fri  ...  34   5.8   0.0   0.00
27   7  4   sep  mon  ...  48   1.3   0.0   0.00
28   6  3   sep  sat  ...  24   2.7   0.0   0.00
29   6  3   sep  sun  ...  39   3.6   0.0   0.00
..  .. ..   ...  ...  ...  ..   ...   ...    ...
487  5  4   aug  tue  ...  34   3.6   0.0  16.40
488  4  4   aug  tue  ...  71   7.6   0.0  46.70
489  4  4   aug  wed  ...  58   1.3   0.0   0.00
490  4  4   aug  wed  ...  33   4.0   0.0   0.00
491  4  4   aug  thu  ...  21   4.5   0.0   0.00
492  1  3   aug  fri  ...  27   2.2   0.0   0.00
493  1  3   aug  fri  ...  29   4.5   0.0  43.32
494  6  6   aug  sat  ...  30   4.9   0.0   8.59
495  6  6   aug  mon  ...  42   2.2   0.0   0.00
496  4  5   aug  mon  ...  26   3.1   0.0   2.77
497  3  4   aug  tue  ...  27   2.2   0.0  14.68
498  6  5   aug  tue  ...  26   2.7   0.0  40.54
499  7  5   aug  tue  ...  63   4.9   6.4  10.82
500  8  6   aug  tue  ...  65   4.9   0.8   0.00
501  7  5   aug  tue  ...  65   4.9   0.8   0.00
502  4  4   aug  tue  ...  69   4.9   0.4   0.00
503  2  4   aug  wed  ...  30   4.9   0.0   1.95
504  4  3   aug  wed  ...  29   4.9   0.0  49.59
505  1  2   aug  thu  ...  35   1.8   0.0   5.80
506  1  2   aug  fri  ...  73   8.5   0.0   0.00
507  2  4   aug  fri  ...  41   3.6   0.0   0.00
508  1  2   aug  fri  ...  41   3.6   0.0   0.00
509  5  4   aug  fri  ...  71   7.6   1.4   2.17
510  6  5   aug  fri  ...  62   5.4   0.0   0.43
511  8  6   aug  sun  ...  35   2.7   0.0   0.00
512  4  3   aug  sun  ...  32   2.7   0.0   6.44
513  2  4   aug  sun  ...  71   5.8   0.0  54.29
514  7  4   aug  sun  ...  70   6.7   0.0  11.16
515  1  4   aug  sat  ...  42   4.0   0.0   0.00
516  6  3   nov  tue  ...  31   4.5   0.0   0.00

[517 rows x 13 columns]

### Oznaczenia poszczególnych kolumn :
  * __X__ - współrzędna osi rzędnych na mapie zamieszczonej poniżej $[1:9]$
  * __Y__ - współrzędna osi odciętych na mapie zamieszczonej poniżej $[2:9]$
  * __month__ - miesiąc wystąpienia pożaru $[jan:dec]$
  * __day__ - dzień wystąpienia pożaru $[mon:sun]$
  * __FFMC__ - Fine Fuel Moisture Code - wartość wilgotności obliczona uwzględniając temperaturę, względną wilgotność powietrza, wiatr i opady deszczu $[{18.7}:{96.2}]$
  * __DMC__ - Duff Moisture Code - współczynnik średniej wilgotności warstw organicznych do średniej głębokości $[{1.1}:{291.3}]$
  * __DC__ - Drought Code - współczynnik wilgotności głębszych warstw organicznych (przydatny przy badaniu obiawów suszy) $[{7.9}:{860.6}]$
  * __ISI__ - Initial Spread Index - współczynnik oczekiwanego rozprzestrzeniania pożaru - łączy wpływy wiatru oraz FFMC $[{0.0}:{56.1}]$
  * __temp__ - temperatura w $C^o$ $[{2.2}:{33.3}]$
  * __RH__ - Relative Hiumidity - wiglotność względna w % $[{15.0}:{100}]$
  * __wind__ - wiatr w km/h $[{0.4}:{40}]$
  * __rain__ - opady deszczu w $\frac{mm}{m^2}$ $[{0.0}:{6.4}]$
  * __area__ - spalona powierzchnia lasu w $ha$ $[{0.0}:{1090.84}]$
  

![img](https://www.researchgate.net/profile/Paulo_Cortez3/publication/238767143/figure/fig1/AS:298804772392991@1448252017812/The-map-of-the-Montesinho-natural-park.png)

<em>Materiał bazowy dostępny na stronie:[http://www.dsi.uminho.pt/~pcortez/fires.pdf](http://www.dsi.uminho.pt/~pcortez/fires.pdf)
    
    

</em>